### PR TITLE
make MQTT subscription tracking thread-safe

### PR DESCRIPTION
**Pull request by J-N-K**
Following  https://github.com/openhab/openhab-core/pull/1259#discussion_r355141114

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

File: bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
 @NonNullByDefault
 public abstract class MqttAsyncClientWrapper {
     private final Logger logger = LoggerFactory.getLogger(MqttAsyncClientWrapper.class);
-    private final List<String> subscribedTopics = new ArrayList<>();
+    private final Set<ClientCallback> subscriptions = new CopyOnWriteArraySet<>();
```
Comment (jochen314) [inline bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java]: I am not sure, if CopyOnWriteArraySet is the best implementation choice here.
All the operations we are doing are add and remove, which are the expensive ones.
I would have gone for <code>ConcurrentHashMap.newKeySet()</code>.

File: bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java [hunk 2]
```diff
@@ -72,23 +72,29 @@
...
+        boolean needsSubscription = subscriptions.add(clientCallback);
+        if (needsSubscription) {
+            logger.trace("Trying to subscribe {} to topic {}", this, topic);
+            return internalSubscribe(topic, qos, clientCallback).thenApply(s -> {
+                logger.trace("Successfully subscribed {} to topic {}", this, topic);
+                return true;
+            });
+        } else {
+            logger.trace("{} already subscribed to topic {}", this, topic);
+            return CompletableFuture.completedFuture(true);
```
Comment (jochen314) [inline bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java]: We could use the result value of the future to indicate, that a real subscription was not needed.
Then we could still do the debug messages in the connection.
Not sure, id that will give us much benefit.

File: bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java]: good point.

File: bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java (hunk 2 above)
Comment (J-N-K) [reply bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java]: I don't think this gives much benefit and currently either true or exceptionally is expected.

**Pull request merged**